### ISSUE TITLE
Hard-to-locate error when providing wrong parameter type to mapState

### ISSUE THREAD

**Issue by joachimboggild** [vuejs/vuex#1162]
### Version
3.0.1

### Reproduction link
[https://codesandbox.io/s/048lw4q59n](https://codesandbox.io/s/048lw4q59n)

### Steps to reproduce
1) Open code sandbox provided.
2) Look inside console to the error message provided.

To fix the problem, include brackets inside the mapState statement on line 97 in the HelloWorld.vue file.

### What is expected?
It was expected that an error message would state that the parameter for the mapState function was invalid in 'HelloWorld.vue'.

### What is actually happening?
This error message is returned: 

```
[Vue warn]: Failed to resolve async component: function HelloWorld() {
  return _promise2.default.resolve().then(function () {
    return require("./components/HelloWorld");
  });
}
Reason: TypeError: Cannot convert undefined or null to object
```

---
I just thought that this was a difficult error to track, because the error message provided did not point towards the actual error. If it is possible to make the error message more precise, it would be great.

**Comment by yawboakye** [comment]
@ktsn Current implementation is confident that either an array or object is always passed, which explains the cryptic error & message: it's not stop earlier. What would you expect to be the behavior when you pass something other than that?

**Comment by EdyHartono** [comment]
#1297

**Comment by sustained** [comment]
Hope that PR makes it in, this has bitten me twice so far.

Maybe I'll learn my lesson (if not after this time then) eventually but it will probably keep biting people.

**Comment by ktsn** [comment]
fixed via #1297
